### PR TITLE
Unicode/UTF-8 Update

### PR DESCRIPTION
**Pull request by smurching**
Updates for better unicode/UTF-8 support in MLflow:

* Assumes text types (`unicode` in Python 2, `str` in Python 3) are used internally in MLflow to represent text values like Project source_name, entry_point, etc. This is consistent with [suggested best practices](https://docs.python.org/3/howto/pyporting.html#text-versus-binary-data) for Python 2/3 compatibility & with [click](http://click.pocoo.org/5/python3/#python-2-and-3-differences) handles text values.
* Uses `yaml.safe_dump` to avoid `!!python/unicode` messages in  YAML files in Python 2.
* Passes `allow_unicode=True` to `yaml.safe_dump` to ensure UTF-8 strings appear properly in on-disk YAML files (as opposed to appearing as ASCII representations of their byte sequences)

### CODE REVIEW

**Comment by mateiz** [comment]
The safe_dump is good, but I don't think we want to allow unicode in metric names, etc necessarily. That said, it seems good to use unicode internally for these.

File: mlflow/cli.py [hunk 1]
```diff
@@ -102,15 +92,15 @@ def run(uri, entry_point, version, param_list, experiment_id, mode, cluster_spec
...
-                     parameters=param_dict, mode=_encode(mode),
-                     cluster_spec=_encode(cluster_spec),
-                     git_username=_encode(git_username),
-                     git_password=_encode(git_password), use_conda=(not no_conda),
-                     use_temp_cwd=new_dir, storage_dir=_encode(storage_dir))
+                     parameters=param_dict, mode=mode,
+                     cluster_spec=cluster_spec,
+                     git_username=git_username,
+                     git_password=git_password, use_conda=(not no_conda),
+                     use_temp_cwd=new_dir, storage_dir=storage_dir)
```
Comment (mateiz) [inline mlflow/cli.py]: Probably better code style to list the parameters one per line?

**Comment by smurching** [comment]
Thanks @mateiz for the review :)

Currently there are still a number of places where either text or binary types could be present in our APIs:
* PyYaml reads text fields as instances of `str`, so run metadata (e.g. source name, source version, etc) are loaded as str. There's not a nice way to override this; gave a [SO suggestion](https://stackoverflow.com/a/2967461) a quick try but it didn't quite work.
* Generated proto code returns `string` proto fields as Python text types, i.e.  as `unicode` in Python 2
* We currently accept both `unicode` and `str` arguments to APIs like `log_metric` & don't apply conversions to text types.

At some point it'd be good to address this, preferably by standardizing around using text types internally, but it could happen later - there's some work towards it in [this branch](https://github.com/smurching/mlflow/tree/utf-8-updates). If you're ok with postponing the full switch to text types I'll go ahead & merge this for now.

**Comment by mateiz** [comment]
Makes sense. I think we should also eventually have a whitelist for allowed characters in metric names, experiment names, etc instead of allowing all of Unicode. For instance, some of them also need to become valid filenames if you use the file store.

**Comment by smurching** [comment]
Yeah that makes sense, plus we could always expand the set of allowed names if there's a need. Thanks for the review, will go ahead & merge this :)

**Pull request merged**